File: src/vpc/runtime/entitlements.js
"use strict";

const { requireUser } = require("../../common/auth");
const { pool } = require("../db/pg");

const ENTITLEMENTS_IMPL = "entitlements-v3";

function numOrNull(v) {
  if (v === null || v === undefined) return null;
  const n = Number(v);
  return Number.isFinite(n) ? n : null;
}

function lower(v) {
  return String(v || "").trim().toLowerCase();
}

async function handleEntitlements({ res, auth }) {
  const okRes = requireUser({ auth, res });
  if (okRes !== true) return okRes;

  const p = pool();
  if (!p) return res.badRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

  const userSub = String(auth.userSub || "").trim();
  if (!userSub) return res.unauthorized("Missing user");

  const nowMs = Date.now();

  // ---------------------------------------
  // v2: RevenueCat-backed premium state first
  // ---------------------------------------
  let rcRow = null;
  let rcErr = null;

  try {
    // IMPORTANT: do NOT swallow errors silently
    const s = await p.query(
      `
      select premium_active, premium_env, product_id, expires_at_ms, last_event_id, last_event_type, last_event_ts_ms
      from user_premium_state
      where app_user_id = $1
      limit 1;
      `,
      [userSub],
    );

    rcRow = s.rows?.[0] || null;
  } catch (e) {
    rcErr = e?.message || String(e);
    console.warn("[entitlements] rc_state query failed:", { impl: ENTITLEMENTS_IMPL, userSub, err: rcErr });
  }

  if (rcRow) {
    const env = lower(rcRow.premium_env);
    const expMs = numOrNull(rcRow.expires_at_ms);
    // prefer expires_at_ms truth if present; otherwise fallback to premium_active
    const active =
      expMs != null ? expMs > nowMs : !!rcRow.premium_active;

    if (active) {
      return res.ok({
        userSub,
        tier: "premium",
        expiresAtMs: expMs,
        unlockedDeckSlugs: [],
        premiumSource: "revenuecat",
        premiumEnv: env || null, // sandbox/production/none
        productId: rcRow.product_id || null,
        serverTimeMs: nowMs,

        // dev diagnostics (safe)
        impl: ENTITLEMENTS_IMPL,
        rcLastEventId: rcRow.last_event_id || null,
        rcLastEventType: rcRow.last_event_type || null,
        rcLastEventTsMs: numOrNull(rcRow.last_event_ts_ms),
      });
    }
  }

  // ---------------------------------------
  // v1 legacy: user_entitlements (deck-based)
  // ---------------------------------------
  const r = await p.query(
    `
    select entitlement_key as "entitlementKey",
           tier,
           expires_at as "expiresAt"
    from user_entitlements
    where user_sub = $1
      and (expires_at is null or expires_at > now())
    order by
      (case when entitlement_key = 'premium_all' then 0 else 1 end) asc,
      expires_at desc nulls first,
      entitlement_key asc
    `,
    [userSub],
  );

  let tier = "free";
  let expiresAtMs = null;
  let unlockedDeckSlugs = [];

  const rows = r.rows || [];

  const premiumAll = rows.find((x) => x.entitlementKey === "premium_all" && x.tier === "premium");
  if (premiumAll) {
    tier = "premium";
    expiresAtMs = premiumAll.expiresAt ? new Date(premiumAll.expiresAt).getTime() : null;
    unlockedDeckSlugs = [];
  } else {
    const deckEnts = rows.filter((x) => String(x.entitlementKey || "").startsWith("deck:") && x.tier === "premium");
    if (deckEnts.length > 0) {
      tier = "premium";
      unlockedDeckSlugs = deckEnts
        .map((x) => String(x.entitlementKey).slice("deck:".length))
        .filter(Boolean);

      const maxExpire = deckEnts
        .map((x) => (x.expiresAt ? new Date(x.expiresAt).getTime() : null))
        .filter((x) => typeof x === "number")
        .sort((a, b) => b - a)[0];
      expiresAtMs = maxExpire ?? null;
    }
  }

  return res.ok({
    userSub,
    tier,
    expiresAtMs,
    unlockedDeckSlugs,
    premiumSource: "legacy",
    premiumEnv: null,
    productId: null,
    serverTimeMs: nowMs,

    // dev diagnostics: only when legacy path is returned
    impl: ENTITLEMENTS_IMPL,
    rcCheckError: rcErr,
    rcRowSeen: !!rcRow,
  });
}

module.exports = { handleEntitlements };

File: src_C/Vpc/Runtime/Entitlements.cs
using System.Globalization;
using Amazon.Lambda.APIGatewayEvents;
using RecallSmith.Lambda.Common;
using RecallSmith.Lambda.Db;

namespace RecallSmith.Lambda.Vpc.Runtime;

public static class Entitlements
{
  private const string EntitlementsImpl = "entitlements-v3";

  public static async Task<APIGatewayProxyResponse> HandleEntitlements(LambdaRequest req, Res res, AuthContext auth)
  {
    var deny = Auth.RequireUser(auth, res);
    if (deny is not null) return deny;

    await using var conn = await Pg.OpenConnectionOrNullAsync();
    if (conn is null) return res.BadRequest("CONFIG_ERROR", "Missing PG env vars (PGHOST/PGDATABASE/PGUSER/PGPASSWORD)");

    var userSub = (auth.UserSub ?? string.Empty).Trim();
    if (string.IsNullOrEmpty(userSub)) return res.Unauthorized("Missing user");

    var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();

    Dictionary<string, object?>? rcRow = null;
    string? rcErr = null;

    try
    {
      const string rcSql = """
        select premium_active, premium_env, product_id, expires_at_ms, last_event_id, last_event_type, last_event_ts_ms
        from user_premium_state
        where app_user_id = $1
        limit 1;
        """;

      var rcRows = await DbUtil.QueryAsync(conn, null, rcSql, [userSub]);
      rcRow = rcRows.Count > 0 ? rcRows[0] : null;
    }
    catch (Exception ex)
    {
      rcErr = ex.Message;
      Log.Warn("[entitlements] rc_state query failed:", new { impl = EntitlementsImpl, userSub, err = rcErr });
    }

    if (rcRow is not null)
    {
      var env = Lower(ToStringOrNull(rcRow.TryGetValue("premium_env", out var pe) ? pe : null));
      var expMs = NumOrNull(rcRow.TryGetValue("expires_at_ms", out var exp) ? exp : null);

      var active = expMs is not null
        ? expMs > nowMs
        : ToBool(rcRow.TryGetValue("premium_active", out var pa) ? pa : null);

      if (active)
      {
        return res.Ok(new
        {
          userSub,
          tier = "premium",
          expiresAtMs = expMs,
          unlockedDeckSlugs = Array.Empty<string>(),
          premiumSource = "revenuecat",
          premiumEnv = string.IsNullOrEmpty(env) ? null : env,
          productId = ToStringOrNull(rcRow.TryGetValue("product_id", out var pid) ? pid : null),
          serverTimeMs = nowMs,

          impl = EntitlementsImpl,
          rcLastEventId = ToStringOrNull(rcRow.TryGetValue("last_event_id", out var le) ? le : null),
          rcLastEventType = ToStringOrNull(rcRow.TryGetValue("last_event_type", out var let) ? let : null),
          rcLastEventTsMs = NumOrNull(rcRow.TryGetValue("last_event_ts_ms", out var lts) ? lts : null),
        });
      }
    }

    // legacy: user_entitlements (deck-based)
    const string legacySql = """
      select entitlement_key as "entitlementKey",
             tier,
             expires_at as "expiresAt"
      from user_entitlements
      where user_sub = $1
        and (expires_at is null or expires_at > now())
      order by
        (case when entitlement_key = 'premium_all' then 0 else 1 end) asc,
        expires_at desc nulls first,
        entitlement_key asc
      """;

    var rows = await DbUtil.QueryAsync(conn, null, legacySql, [userSub]);

    var tierOut = "free";
    long? expiresAtMs = null;
    var unlockedDeckSlugs = new List<string>();

    var premiumAll = rows.FirstOrDefault(x =>
      string.Equals(Convert.ToString(x.GetValueOrDefault("entitlementKey")), "premium_all", StringComparison.Ordinal) &&
      string.Equals(Convert.ToString(x.GetValueOrDefault("tier")), "premium", StringComparison.Ordinal));

    if (premiumAll is not null)
    {
      tierOut = "premium";
      expiresAtMs = DateToMs(premiumAll.GetValueOrDefault("expiresAt"));
    }
    else
    {
      var deckEnts = rows.Where(x =>
        Convert.ToString(x.GetValueOrDefault("tier")) == "premium" &&
        (Convert.ToString(x.GetValueOrDefault("entitlementKey")) ?? string.Empty).StartsWith("deck:", StringComparison.Ordinal)).ToList();

      if (deckEnts.Count > 0)
      {
        tierOut = "premium";
        unlockedDeckSlugs = deckEnts
          .Select(x => Convert.ToString(x.GetValueOrDefault("entitlementKey")) ?? string.Empty)
          .Select(k => k.Length > "deck:".Length ? k["deck:".Length..] : string.Empty)
          .Where(s => s.Length > 0)
          .ToList();

        expiresAtMs = deckEnts
          .Select(x => DateToMs(x.GetValueOrDefault("expiresAt")))
          .Where(x => x is not null)
          .OrderByDescending(x => x)
          .FirstOrDefault();
      }
    }

    return res.Ok(new
    {
      userSub,
      tier = tierOut,
      expiresAtMs,
      unlockedDeckSlugs,
      premiumSource = "legacy",
      premiumEnv = (string?)null,
      productId = (string?)null,
      serverTimeMs = nowMs,

      impl = EntitlementsImpl,
      rcCheckError = rcErr,
      rcRowSeen = rcRow is not null,
    });
  }

  private static string? ToStringOrNull(object? v)
  {
    var s = Convert.ToString(v, CultureInfo.InvariantCulture);
    return string.IsNullOrEmpty(s) ? null : s;
  }

  private static string Lower(string? v) => (v ?? string.Empty).Trim().ToLowerInvariant();

  private static long? NumOrNull(object? v)
  {
    if (v is null) return null;
    try
    {
      return v switch
      {
        long l => l,
        int i => i,
        short s => s,
        decimal d => (long)d,
        double d => (long)Math.Floor(d),
        float f => (long)Math.Floor(f),
        _ => long.TryParse(Convert.ToString(v, CultureInfo.InvariantCulture), NumberStyles.Integer, CultureInfo.InvariantCulture, out var n) ? n : null,
      };
    }
    catch
    {
      return null;
    }
  }

  private static bool ToBool(object? v)
  {
    if (v is null) return false;
    if (v is bool b) return b;
    var s = Convert.ToString(v, CultureInfo.InvariantCulture)?.Trim().ToLowerInvariant();
    return s is "1" or "true" or "yes" or "y";
  }

  private static long? DateToMs(object? v)
  {
    if (v is null) return null;
    if (v is DateTime dt) return new DateTimeOffset(dt).ToUnixTimeMilliseconds();
    if (v is DateTimeOffset dto) return dto.ToUnixTimeMilliseconds();
    return null;
  }

  private static object? GetValueOrDefault(this Dictionary<string, object?> dict, string key)
  {
    return dict.TryGetValue(key, out var v) ? v : null;
  }
}

